### PR TITLE
GH-43060: [C++][Parquet] Optimize memory usage in buffering parquet::ReaderProperties::GetStream

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

See https://github.com/apache/arrow/issues/43060 . This patch optimize memory-usage for column is not so large and buffering is enabled.

### What changes are included in this PR?

Change buffer-size in `parquet::ReaderProperties::GetStream` and making it not greater than column-chunk size.

### Are these changes tested?

Covered by existing

### Are there any user-facing changes?

No

* GitHub Issue: #43060

### CODE REVIEW

File: cpp/src/parquet/properties.cc [hunk 1]
```diff
@@ -34,10 +34,10 @@ std::shared_ptr<ArrowInputStream> ReaderProperties::GetStream(
     // of source
     PARQUET_ASSIGN_OR_THROW(
         std::shared_ptr<::arrow::io::InputStream> safe_stream,
-        ::arrow::io::RandomAccessFile::GetStream(source, start, num_bytes));
-    PARQUET_ASSIGN_OR_THROW(
-        auto stream, ::arrow::io::BufferedInputStream::Create(buffer_size_, pool_,
-                                                              safe_stream, num_bytes));
+        ::arrow::io::RandomAccessFile::GetStream(std::move(source), start, num_bytes));
+    PARQUET_ASSIGN_OR_THROW(auto stream, ::arrow::io::BufferedInputStream::Create(
+                                             std::min(buffer_size_, num_bytes), pool_,
```
Comment (mapleFU) [inline cpp/src/parquet/properties.cc:39]: I don't know should I add a `std::max(std::min(buffer_size_, num_bytes), 1U)`...

**Comment by pitrou** [comment]
How about instead handling this in `BufferedInputStream::SetBufferSize`? Something like (untested):
```diff
diff --git a/cpp/src/arrow/io/buffered.cc b/cpp/src/arrow/io/buffered.cc
index c53b3d223d..18d1009427 100644
--- a/cpp/src/arrow/io/buffered.cc
+++ b/cpp/src/arrow/io/buffered.cc
@@ -286,6 +286,11 @@ class BufferedInputStream::Impl : public BufferedBase {
     if (new_buffer_size <= 0) {
       return Status::Invalid("Buffer size should be positive");
     }
+    if (raw_read_bound_ >= 0) {
+      // No need to reserve space for more than the total remaining number of bytes.
+      new_buffer_size = std::min(new_buffer_size,
+                                 bytes_buffered_ + (raw_read_bound_ - raw_read_total_));
+    }
     if ((buffer_pos_ + bytes_buffered_) >= new_buffer_size) {
       return Status::Invalid("Cannot shrink read buffer if buffered data remains");
     }
```

**Comment by mapleFU** [comment]
They both LGTM. This `parquet::ReaderProperties::GetStream` is used for each column chunk ( with known io-range). Handling this in `BufferedInputStream::SetBufferSize` is also LGTM and have same syntax here. ( Though final buffer of `BufferedInputStream` might also use more memory than expected, I think that scenerio doesn't matters me now)

**Comment by mapleFU** [comment]
I'll draft another pr for BufferedInput, you can decide which to merge. These two are both LGTM

**Comment by pitrou** [comment]
Doing it in `BufferedInputStream::SetBufferSize` would address all uses of `BufferedInputStream` instead of making it a Parquet-specific fix.

**Comment by mapleFU** [comment]
I've issue another pr here: https://github.com/apache/arrow/pull/43064

**Pull request closed without merging**